src/landscape.cpp

```diff
@@ -1321,75 +1323,72 @@ static void BuildRiver(TileIndex begin, TileIndex end, TileIndex spring, bool ma
  */
 static std::tuple<bool, bool> FlowRiver(TileIndex spring, TileIndex begin, uint min_river_length)
 {
-	uint height_begin = TileHeight(begin);
-
 	if (IsWaterTile(begin)) {
 		return { DistanceManhattan(spring, begin) > min_river_length, GetTileZ(begin) == 0 };
 	}
 
-	std::set<TileIndex> marks;
-	marks.insert(begin);
+	int height_begin = TileHeight(begin);
 
-	/* Breadth first search for the closest tile we can flow down to. */
-	std::list<TileIndex> queue;
+	static std::unordered_set<TileIndex::BaseType> marks;
+	marks.clear();
+	marks.insert(Tile(begin));
+
+	static std::vector<TileIndex> queue;
+	queue.clear();
 	queue.push_back(begin);
 
+	/* Breadth first search for the closest tile we can flow down to. */
 	bool found = false;
-	uint count = 0; // Number of tiles considered; to be used for lake location guessing.
 	TileIndex end;
-	do {
-		end = queue.front();
-		queue.pop_front();
+	for (size_t i = 0; i != queue.size(); i++) {
+		end = queue[i];
```

Comment:
This doesn't really feel right. Why retain the queue that you already processed? If `std::list` is slow, maybe use `std::deque` and keep the code as it was.

Reply:
I keep the queue to find the n-th tile for lake guessing. The queue is neatly ordered by insertion.